### DISCUSSION TITLE
HOC goes against react and is making this plugin hard to use

### DISCUSSION THREAD

**Discussion by jnovak-SM2Dev** [dowjones/react-dropdown-tree-select#602] [General]
To bring this up again, I think the HOC fix is not a great way to go for this plugin. I'm using this for a select that shows how many of the options were found in the search, so the label of each item will change as the user selects options. This obviously causes the entire thing to re-render (even with hoc) since the data changed.

I think fighting against the natural way react works is making this component a lot harder to use than it needs to be. If this just had a "selected" or "values" field that could be updated and would find and check said values in the data, I think it would work a lot better and be easier to use. I see one of the arguments was that the plugin should maintain it's own state, but that isn't how react components normally work and is not how forms generally work in react. Especially when you are passing in the data to the component that controls what it shows and when that data updates the component loses all semblance of what was open. That to me is a huge problem and is the opposite of how any other select component out there works.

**Comment by gandhis1** [comment]
Can you elaborate on why you believe this "goes against React"? React is built on using re-renders for changed state.

**Reply by jnovak-SM2Dev** [reply]
As I mentioned, when you update the data for this plugin the component re-renders and loses track of what the user had open. This makes the component hard to use. The "work around" is to basically make the component not re-render, which defeats the purpose of react and only works if the data actually did not change.

For example:

1. The data is set. The user checks some values. The data changes, but you supress the change by filtering out selected values and using HOC.
2. The data is set. The user checks some values. The labels need to be updated. The component re-renders even with HOC because the component data did change. The select re-renders while still open and to the user all the trees collapse and the user has to start over.

**Reply by jnovak-SM2Dev** [reply]
If the select had an "selected" or "values" field. You could update that when a value was changed and the component itself would only need to keep track of what options were open (which it could do by value). I think this would be more inline with how every other react select component works and would prevent 40 people from asking why this component breaks on state change.

**Reply by mrchief** [reply]
It is trickier than you think and this component is not like every other react component. Very few handle large data nodes and even fewer that are able to handle thousands of nodes while not bogging the page down. There are trade offs but this component making up for basic lack of understanding of how the framework you're using is a bigger anti-pattern. 

I'm open to suggestions and if you have ideas about solving this then please send a PR.
